wallet-new/test/unit/Test/Spec/Kernel.hs

```diff
@@ -67,17 +80,66 @@ spec =
                     => Kernel.ActiveWallet
                     -> Inductive h Addr
                     -> Expectation
-    checkEquivalent activeWallet ind = do
-       shouldReturnValidated $ runTranslateTNoErrors $ do
+    checkEquivalent w ind = shouldReturnValidated $ evaluate w ind
+
+    -- | Evaluate the inductive wallet step by step and compare the DSL and Cardano results
+    --   at the end of each step.
+    -- NOTE: This evaluation changes the state of the wallet and also produces an
+    -- interpretation context, which we return to enable further custom interpretation
+    evaluate :: forall h. Hash h Addr
+             => Kernel.ActiveWallet
+             -> Inductive h Addr
+             -> IO (Validated EquivalenceViolation (IntCtxt h))
+    evaluate activeWallet ind = do
+       fmap (fmap snd) $ runTranslateTNoErrors $ do
          equivalentT activeWallet (encKpEnc ekp) (mkWallet (== addr)) ind
       where
         [addr]       = Set.toList $ inductiveOurs ind
         AddrInfo{..} = resolveAddr addr transCtxt
         Just ekp     = addrInfoMasterKey
 
+    evaluate' :: forall h. Hash h Addr
+             => Kernel.ActiveWallet
+             -> Inductive h Addr
+             -> IO (IntCtxt h)
+    evaluate' activeWallet ind = do
+        res <- evaluate activeWallet ind
+        case res of
+            Invalid _ e    -> throwM e
+            Valid intCtxt' -> return intCtxt'
+
     mkWallet :: Hash h Addr => Ours Addr -> Transaction h Addr -> Wallet h Addr
     mkWallet = walletBoot Full.walletEmpty
 
+    -- | Translates the DSL BlockMeta' value to BlockMeta
+    intBlockMeta :: forall h. (Hash h Addr, Interpret h (BlockMeta' h))
+                 => IntCtxt h
+                 -> (BlockMeta' h)
+                 -> TranslateT IntException IO BlockMeta
+    intBlockMeta intCtxt a = do
+        ma' <- catchTranslateErrors $ runIntT' intCtxt $ int a
+        case ma' of
+          Left err -> liftIO $ throwM err
```

Comment:
Is the intent to throw the exception in `IO`? It seems like `TranslateT` has some means of carrying exception/error information.

Reply:
I guess the point is to have the interpretation exception be thrown up to the test framework, and have the test marked as failed should that happen.

Reply:
Sounds good :)